### ISSUE TITLE
 deadlock in pkg/kvstore/lock.go:87 kvstore.LockPath { ll.Lock() } <<<<<

### ISSUE THREAD

**Issue by aalemayhu** [cilium/cilium#2624]
While testing #2607 noticed the below deadlock. Ran the runtime tests a couple of times to trigger. See attached log file for full output. 

```
sh-4.4$ git ll
5ccf490fcc99 (api: Include L4 rule labels in endpoint info, 2017-12-15)
sh-4.4$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                                                                NAMES
418192398f67        consul:0.8.3        "docker-entrypoint..."   About an hour ago   Up About an hour    8300-8302/tcp, 8400/tcp, 8301-8302/udp, 8600/tcp, 8600/udp, 0.0.0.0:8500->8500/tcp   cilium-consul
sh-4.4$ journalctl -S today -u cilium | grep DEADLOCK -F10
Jan 24 10:48:07 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc420297200 Labels:[]}"
Jan 24 10:48:08 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc42013ed00 Labels:[]}"
Jan 24 10:48:09 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc420297700 Labels:[]}"
Jan 24 10:48:10 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc420e68c00 Labels:[]}"
Jan 24 10:48:11 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc4205a5500 Labels:[]}"
Jan 24 10:48:12 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc4205a5700 Labels:[]}"
Jan 24 10:48:13 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc42013f500 Labels:[]}"
Jan 24 10:48:14 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc420cee300 Labels:[]}"
Jan 24 10:48:15 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc42013f800 Labels:[]}"
Jan 24 10:48:16 anam cilium-agent[4731]: level=debug msg="GET /endpoint request" params="{HTTPRequest:0xc420995000 Labels:[]}"
Jan 24 10:48:17 anam cilium-agent[4731]: POTENTIAL DEADLOCK:
Jan 24 10:48:17 anam cilium-agent[4731]: Previous place where the lock was grabbed
Jan 24 10:48:17 anam cilium-agent[4731]: goroutine 4051 lock 0xc420c77028
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/kvstore/lock.go:87 kvstore.LockPath { ll.Lock() } <<<<<
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/daemon/labels.go:317 main.(*Daemon).DeleteIdentityBySHA256 { lockKey, err := kvstore.LockPath(lblPath) }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/daemon/endpoint.go:358 main.(*Daemon).deleteEndpoint { if err := d.DeleteIdentityBySHA256(sha256sum, ep.StringID()); err != nil { }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/daemon/endpoint.go:435 main.(*Daemon).DeleteEndpoint { return d.deleteEndpoint(ep), nil }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/daemon/endpoint.go:451 main.(*deleteEndpointID).Handle { if nerr, err := d.DeleteEndpoint(params.ID); err != nil { }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/api/v1/server/restapi/endpoint/delete_endpoint_id.go:65 endpoint.(*DeleteEndpointID).ServeHTTP { res := o.Handler.Handle(Params) // actually handle the request }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/og/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 middleware.NewOperationExecutor.func1 { route.Handler.ServeHTTP(rw, r) }
Jan 24 10:48:17 anam cilium-agent[4731]: /home/scanf/opt/go/src/net/http/server.go:1918 http.HandlerFunc.ServeHTTP { f(w, r) }
```
[heute.txt](https://github.com/cilium/cilium/files/1659437/heute.txt)

**Comment by aalemayhu** [comment]
Forgot to add that race and deadlock detection has to be enabled
```diff
diff --git a/Makefile.defs b/Makefile.defs
index ddd103e2e2c3..ebadf378a644 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -17,10 +17,10 @@ BUILD = $(VERSION) $(GIT_VERSION) $(shell go version)
 GOBUILD = -ldflags '-X "github.com/cilium/cilium/pkg/version.Version=$(BUILD)"'
 
 # Uncomment to enable race detection
-#GOBUILD += -race
+GOBUILD += -race
 
 # Uncomment to enable deadlock detection
-#GOBUILD += -tags lockdebug
+GOBUILD += -tags lockdebug
 
 ifneq ($(LOCKDEBUG),)
     GOBUILD += -tags lockdebug
```

**Comment by tgraf** [comment]
Will look at this as part of #2613

**Comment by tgraf** [comment]
Fixed by #2708
